warehouse/query-core/src/main/java/datawave/query/jexl/visitors/TreeFlatteningRebuildingVisitor.java

```diff
@@ -70,44 +72,78 @@ public Object visit(ASTOrNode node, Object data) {
         ASTOrNode orNode = JexlNodes.newInstanceOfType(node);
         orNode.image = node.image;
         
-        return flattenTree(node, orNode, data);
+        LinkedList<JexlNode> stackA = new LinkedList<>();
+        stackA.push(node);
+        LinkedList<JexlNode> stackB;
+        boolean hasMoreOrNodes;
+        do {
+            stackB = new LinkedList<>();
+            hasMoreOrNodes = false;
+            do {
+                JexlNode currentNode = stackA.pop();
+                if (currentNode instanceof ASTOrNode) {
+                    for (int i = currentNode.jjtGetNumChildren() - 1; i >= 0; i--) {
+                        JexlNode child = currentNode.jjtGetChild(i);
+                        JexlNode dereferenced = JexlASTHelper.dereference(child);
+                        if (dereferenced instanceof ASTOrNode) {
+                            hasMoreOrNodes = true;
+                            stackB.push(dereferenced);
+                        } else {
+                            stackB.push(child);
+                        }
+                    }
+                } else {
+                    stackB.push(currentNode);
+                }
+            } while (!stackA.isEmpty());
+            stackA = new LinkedList<>(stackB);
```

Comment:
I'm worried about the efficiency of this implementation.  The way this is written, you are iterating over all of your nodes (even the ones you know aren't OR nodes) repeatedly until you complete an iteration without encountering an OR node.  Would it not be better to separate out the known non-OR nodes into a separate stack as you work?

The implementation I shared with you would avoid the repeated traversals over nodes which you already know are not OR nodes, and would save us time during the query planning phase (especially since flatten tends to get called multiple times during planning).  It might be good to add a test which iterates over an OR node of 100000+ nodes to get a feel for the performance.